proc/process.c

```diff
@@ -877,93 +874,80 @@ int process_load(process_t *process, vm_object_t *o, off_t base, size_t size, vo
 		return -ENOEXEC;
 	}
 
-#ifdef __sparc__
 	/* find symtab */
 	for (i = 0, shdr = (void *)((char *)ehdr + ehdr->e_shoff); i < ehdr->e_shnum; i++, shdr++) {
 		if (hal_strcmp(&snameTab[shdr->sh_name], ".symtab") == 0) {
+			symTab = (ptr_t)ehdr + (ptr_t)shdr->sh_offset;
 			break;
 		}
 	}
 
-	if (i >= ehdr->e_shnum) {
-		return -ENOEXEC;
-	}
-	symTab = (ptr_t)ehdr + (ptr_t)shdr->sh_offset;
-
 	/* Perform data, init_array and fini_array relocation */
 	for (i = 0, shdr = (void *)((char *)ehdr + ehdr->e_shoff); i < ehdr->e_shnum; i++, shdr++) {
-		/* strncmp as there may be multiple .rela.* sections for different sections.  */
-		if (hal_strncmp(&snameTab[shdr->sh_name], ".rela", 5) != 0) {
-			continue;
-		}
-
 		if ((shdr->sh_size == 0) || (shdr->sh_entsize == 0)) {
 			continue;
 		}
 
-		for (j = 0; j < shdr->sh_size / shdr->sh_entsize; ++j) {
-			rela = (Elf32_Rela *)((ptr_t)shdr->sh_offset + (ptr_t)ehdr + (j * shdr->sh_entsize));
-			reltype = ELF32_R_TYPE(rela->r_info);
-
-			if (reltype == R_SPARC_32) {
-				relptr = (void *)rela->r_offset;
-				if (process_relocate(reloc, relocsz, (char **)&relptr) < 0) {
-					return -ENOEXEC;
-				}
-
-				/* Don't modify ELF file! */
-				if (((ptr_t)relptr >= (ptr_t)base) && ((ptr_t)relptr < ((ptr_t)base + size))) {
-					++badreloc;
-					continue;
-				}
-
-				sym = (Elf32_Sym *)(symTab + (ELF32_R_SYM(rela->r_info) * sizeof(Elf32_Sym)));
-
-				/* Write addend to the address */
-				*(char **)relptr = (char *)(sym->st_value + rela->r_addend);
-
-				if (process_relocate(reloc, relocsz, relptr) < 0) {
-					return -ENOEXEC;
+		/* strncmp as there may be multiple .rela.* or .rel.* sections for different sections.  */
+		if (hal_strncmp(&snameTab[shdr->sh_name], ".rela.", 6) == 0) {
+			/* From currently supported NOMMU platforms only SPARC uses RELA */
+			for (j = 0; j < shdr->sh_size / shdr->sh_entsize; ++j) {
+				rela = (Elf32_Rela *)((ptr_t)shdr->sh_offset + (ptr_t)ehdr + (j * shdr->sh_entsize));
+				reltype = ELF32_R_TYPE(rela->r_info);
+
+				if (reltype == R_SPARC_32) {
+					if (symTab == NULL) {
+						return -ENOEXEC;
+					}
+
+					relptr = (void *)rela->r_offset;
+					if (process_relocate(reloc, relocsz, (char **)&relptr) < 0) {
+						return -ENOEXEC;
+					}
+
+					/* Don't modify ELF file! */
+					if (((ptr_t)relptr >= (ptr_t)base) && ((ptr_t)relptr < ((ptr_t)base + size))) {
+						++badreloc;
+						continue;
+					}
+
+					sym = (Elf32_Sym *)(symTab + (ELF32_R_SYM(rela->r_info) * sizeof(Elf32_Sym)));
+
+					/* Write addend to the address */
+					*(char **)relptr = (char *)(sym->st_value + rela->r_addend);
```

Comment:
This is a pretty much the only difference between SPARC and ARM stuff. Perhaps we could store strcmp result and use it here to do this operation conditionally?

Reply:
I thought about it and it was a bunny hole.

Each architecture has a separate set of Relocation Types, however relocation types are just numbers and there maybe a clash eg. `R_SPARC_32` on SPARC  and `R_ARM_REL32`(unused now) on ARM have the same value of 3. Thus, we should compare relocation type separately on each architecture. This can be done either via ifdefs or via checking if in the ELF header the machine type(`e_machine` field). Now this clash will not happen, as SPARC uses RELA and ARM uses REL. If we unify the code there is high potential for a bug in the future if more relocation types would be needed for some reason.

In my opinion the code should be modified so that:
1) Getting `rela` section on an architecture using `rel` would result in `ENOEXEC` and vice versa
2) We should have separate relocations handler for each architecture

What do you think about it? Also, Should we check the architecture via ifdef or the ELF HEADER?

Reply:
Well, first question that has to be asked is if we'll ever need to handle any more relocation types in the kernel? Aren't dynamic linker solely in the userspace?

Regarding collisions of values of relocation types between archs - this can be solved by e.g. defining our own, abstract, machine independent relocation types on per arch basis, and handle these instead of e.g. `R_SPARC_32` directly (that might be represented by e.g. `HAL_R_REL32` or something along these lines).

Reply:
1) To support static FDPIC program some 2 more relocations are needed, unless we want to move relocation process to the userspace. 

Additionally if we ever want to handle C++ on NOMMU then to have proper exceptions handling there may be a need to add few more relocations. However that's a harder topic as currently exception handling causes relocations in text segment :skull_and_crossbones: 

2) Nice idea!

Also my bad, both ARM and SPARC support in ABI both rela and rel, however GCC by default uses only one of those types if possible.

Reply:
My perspective is as follows - better for this code to do the thing it has to do nice and simple and to not prepare for the future that might never come (kinda a part of the top-down philosophy).

From my side - the proposed version is way better and status quo as is, so I have no objection to merge if you wish to not make any further changes.

Reply:
I checked and compilation with `-static-pie` fixes issue of relocation in text segment in exceptions handling section

Reply:
I have adapted code to you suggestions, and added feature to handle both Rel and Rela relocations on both targets